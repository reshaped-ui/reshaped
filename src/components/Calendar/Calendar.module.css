.selection {
	table-layout: fixed;
	width: 100%;
}

.weekday {
	color: var(--rs-color-foreground-neutral-faded);
	font-weight: var(--rs-font-weight-regular);
	padding-bottom: var(--rs-unit-x2);
}

.row {
	border-top: 2px solid transparent;
}

[dir="rtl"] .control {
	transform: scaleX(-1);
}

.cell-button {
	padding: var(--rs-unit-x2);
	transition: var(--rs-duration-fast) var(--rs-easing-standard);
	transition-property: border-radius, background-color, box-shadow, color;
	text-align: center;

	&[disabled],
	&[disabled]:hover {
		background-color: transparent !important;
		color: var(--rs-color-foreground-disabled) !important;
	}
}

.cell {
	padding: 0;

	& .cell-button:hover,
	&--in-range .cell-button {
		background-color: rgba(var(--rs-color-rgb-background-neutral), 32%);
	}

	&--active-start .cell-button,
	&--active-end .cell-button,
	&--active-start:hover .cell-button,
	&--active-end:hover .cell-button {
		background-color: var(--rs-color-background-primary);
		color: var(--rs-color-on-background-primary);
	}

	/* First day of the week */
	&:first-child .cell-button,
	&--active-start .cell-button {
		border-start-start-radius: var(--rs-radius-small);
		border-end-start-radius: var(--rs-radius-small);
	}

	/* Last day of the week */
	&:last-child .cell-button,
	&--active-end .cell-button,
	/* Hovered date after start date or the highlighted range */
	&--in-range + &:hover:not(&--in-range) .cell-button,
	&--active-start + &:hover:not(&--in-range) .cell-button {
		border-start-end-radius: var(--rs-radius-small);
		border-end-end-radius: var(--rs-radius-small);
	}

	/* Hovered date outside the selection and range highlighted dates */
	&:hover:not(&--in-range, &--in-range + &:hover, &--active-start + &:hover) .cell-button,
	&--active-single + &:hover .cell-button {
		border-radius: var(--rs-radius-small);
	}
}
