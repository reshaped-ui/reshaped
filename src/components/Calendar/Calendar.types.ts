export type BaseProps = {
	defaultMonth?: Date;
	min?: Date;
	max?: Date;
};

type ControlledSingleProps = {
	value: Date | null;
	defaultValue?: never;
	onChange?: (args: { value: Date }) => void;
};

type UncontrolledSingleProps = {
	value?: never;
	defaultValue?: Date;
	onChange?: (args: { value: Date }) => void;
};

export type UncontrolledProps = UncontrolledSingleProps;
export type ControlledProps = ControlledSingleProps;
export type SingleProps = ControlledSingleProps | UncontrolledSingleProps;

export type Props = SingleProps & BaseProps;

export type MonthProps = {
	date: Date;
} & ControlledProps &
	Pick<BaseProps, "max" | "min">;

export type DateProps = {
	date: Date | null;
	active?: boolean;
	disabled?: boolean;
	onClick?: () => void;
};
