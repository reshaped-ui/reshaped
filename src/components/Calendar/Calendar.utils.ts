const DAYS_IN_WEEK = 7;
const FIRST_WEEKDAY = 1;

/**
 * Return the ISO date format without the timezones adjustemnts
 */
export const getLocalISODate = (args: { date: Date }) => {
	const { date } = args;

	return [
		date.getFullYear(),
		(date.getMonth() + 1).toString().padStart(2),
		date.getDate().toString().padStart(2),
	].join("-");
};

const getNormalizedDay = (args: { date: Date }) => {
	const { date } = args;

	const day = date.getDay();
	return day < FIRST_WEEKDAY ? DAYS_IN_WEEK - day - FIRST_WEEKDAY : day - FIRST_WEEKDAY;
};

/**
 * Return an array of US weekday names for the calendar
 */
export const getWeekdayNames = () => {
	const baseDate = new Date(Date.UTC(2021, 0, 4)); // Starting from a Monday
	const weekdays = [];

	for (let i = 0; i < 7; i++) {
		const weekday = baseDate.toLocaleDateString("en-US", { weekday: "short" });
		weekdays.push(weekday.slice(0, 2));
		baseDate.setDate(baseDate.getDate() + 1);
	}

	return weekdays;
};

/**
 * Return an array of all month names
 */
export const getMonthNames = () => {
	return new Array(12).fill(null).map((_, i) => {
		const date = new Date(0, i);
		return date.toLocaleString("default", { month: "short" });
	});
};

/**
 * Return an array of weeks based on the month passed to the function
 */
export const getMonthWeeks = (args: { date: Date }) => {
	const { date } = args;
	const month = date.getMonth();
	const year = date.getFullYear();
	const weeks: Date[][] = [];
	const currentDate = new Date(year, month, 1);

	// Fill in the days if month starts in the middle of the week
	const firstDay = getNormalizedDay({ date: currentDate });
	if (firstDay !== 0) weeks.push(new Array(firstDay).fill(null));

	while (month === currentDate.getMonth()) {
		const day = getNormalizedDay({ date: currentDate });

		if (day === 0 || !weeks.length) weeks.push([]);

		weeks[weeks.length - 1].push(new Date(currentDate));
		currentDate.setDate(currentDate.getDate() + 1);
	}

	// Fill in the days if month ends in the middle of the week
	const lastDay = getNormalizedDay({ date: currentDate });
	if (lastDay !== 0) weeks[weeks.length - 1].push(...new Array(7 - lastDay).fill(null));

	return weeks;
};

export const setMonthTo = (date: Date, value: number) => {
	const resultDate = new Date(date);

	resultDate.setMonth(value);
	return resultDate;
};

export const setMonthToPrevious = (date: Date) => {
	return setMonthTo(date, date.getMonth() - 1);
};

export const setMonthToNext = (date: Date) => {
	return setMonthTo(date, date.getMonth() + 1);
};

export const setYearTo = (date: Date, value: number) => {
	const resultDate = new Date(date);

	resultDate.setFullYear(value);
	return resultDate;
};

export const setYearToPrevious = (date: Date) => {
	return setYearTo(date, date.getFullYear() - 1);
};

export const setYearToNext = (date: Date) => {
	return setYearTo(date, date.getFullYear() + 1);
};

export const applyNavigationBounds = (args: { date: Date; min?: Date; max?: Date }) => {
	const { date, min, max } = args;
	const currentMonth = date.getMonth();
	const currentYear = date.getFullYear();
	const prevMonthLastDate = new Date(currentYear, currentMonth, 0);
	const nextMonthFirstDate = setMonthToNext(date);
	nextMonthFirstDate.setDate(0);

	return {
		isFirstMonth: min && min > prevMonthLastDate,
		isLastMonth: max && max < nextMonthFirstDate,
	};
};
