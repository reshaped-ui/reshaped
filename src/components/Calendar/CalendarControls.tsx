import React from "react";
import Button from "components/Button";
import View from "components/View";
import Hidden from "components/Hidden";
import Text from "components/Text";
import HiddenVisually from "components/HiddenVisually";
import IconChevronLeft from "icons/ChevronLeft";
import IconChevronRight from "icons/ChevronRight";
import { onNextFrame } from "utilities/animation";
import type * as T from "./Calendar.types";
import s from "./Calendar.module.css";

const CalendarControls = (props: T.ControlsProps) => {
	const {
		selectionMode,
		onMonthTitleClick,
		monthTitleRef,
		monthDate,
		renderSelectedMonthLabel,
		isFirstMonth,
		isLastMonth,
		onNextClick,
		onPreviousClick,
		monthSelectionAriaLabel = "Select a month",
		previousMonthAriaLabel = "Previous month",
		previousYearAriaLabel = "Previous year",
		nextMonthAriaLabel = "Next month",
		nextYearAriaLabel = "Next year",
	} = props;
	const prevRef = React.useRef<HTMLButtonElement>(null);
	const nextRef = React.useRef<HTMLButtonElement>(null);

	React.useEffect(() => {
		if (!isFirstMonth) return;
		if (document.activeElement !== prevRef.current) return;

		const targetEl = nextRef.current || monthTitleRef.current;

		onNextFrame(() => {
			targetEl?.focus();
		});
	}, [isFirstMonth, monthTitleRef]);

	React.useEffect(() => {
		if (!isLastMonth) return;
		if (document.activeElement !== nextRef.current) return;

		const targetEl = prevRef.current || monthTitleRef.current;

		onNextFrame(() => {
			targetEl?.focus();
		});
	}, [isLastMonth, monthTitleRef]);

	return (
		<View direction="row" gap={2} align="center">
			<Hidden visibility hide={isFirstMonth}>
				<div className={s.control}>
					<Button
						variant="ghost"
						icon={IconChevronLeft}
						onClick={onPreviousClick}
						attributes={{
							ref: prevRef,
							"aria-label":
								selectionMode === "date" ? previousMonthAriaLabel : previousYearAriaLabel,
						}}
					/>
				</div>
			</Hidden>
			<View.Item grow>
				{selectionMode === "date" && (
					<Button
						fullWidth
						variant="ghost"
						onClick={onMonthTitleClick}
						attributes={{ ref: monthTitleRef }}
					>
						{renderSelectedMonthLabel
							? renderSelectedMonthLabel({ date: monthDate })
							: monthDate.toLocaleDateString("en-US", { month: "long", year: "numeric" })}
						<HiddenVisually>{monthSelectionAriaLabel}</HiddenVisually>
					</Button>
				)}

				{selectionMode === "month" && (
					<Text align="center" weight="medium">
						{monthDate.toLocaleDateString("en-US", { year: "numeric" })}
					</Text>
				)}
			</View.Item>
			<Hidden visibility hide={isLastMonth}>
				<div className={s.control}>
					<Button
						variant="ghost"
						icon={IconChevronRight}
						onClick={onNextClick}
						attributes={{
							ref: nextRef,
							"aria-label": selectionMode === "date" ? nextMonthAriaLabel : nextYearAriaLabel,
						}}
					/>
				</div>
			</Hidden>
		</View>
	);
};

CalendarControls.displayName = "CalendarControls";

export default CalendarControls;
