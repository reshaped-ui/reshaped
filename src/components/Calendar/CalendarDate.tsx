"use client";

import React from "react";
import Actionable from "components/Actionable";
import { classNames } from "utilities/helpers";
import { getLocalISODate } from "./Calendar.utils";
import type * as T from "./Calendar.types";
import s from "./Calendar.module.css";

const CalendarDate = (props: T.DateProps) => {
	const {
		date,
		startValue,
		endValue,
		isActiveStart,
		isActiveEnd,
		disabled,
		focusable,
		onChange,
		range,
		hoveredDate,
		onDateHover,
		onDateHoverEnd,
		renderAriaLabel,
	} = props;

	if (!date) return <td className={s.cell} aria-hidden="true" />;

	const inRange =
		startValue &&
		startValue < date &&
		((endValue && endValue > date) || (hoveredDate && !endValue && hoveredDate > date));
	const dateClassNames = classNames([
		s.cell,
		isActiveStart && s["cell--active-start"],
		(isActiveEnd ||
			(!endValue && isActiveStart && !(hoveredDate && startValue && hoveredDate > startValue))) &&
			s["cell--active-end"],
		inRange && s["cell--in-range"],
	]);

	const handleClick = () => {
		if (!range) {
			(onChange as T.SingleProps["onChange"])?.({ value: date });
			return;
		}

		const bothDatesSelected = startValue && endValue;
		const noneDatesSelected = !startValue && !endValue;
		const selectingBeforeStart = startValue && date < startValue;
		const selectingStart = bothDatesSelected || noneDatesSelected || selectingBeforeStart;
		const start = selectingStart ? date : startValue;
		const end = !selectingStart ? date : null;

		(onChange as T.RangeProps["onChange"])?.({ value: { start, end } });
	};

	const handleMouseEnter = () => {
		onDateHover(date);
	};

	const handleMouseLeave = () => {
		onDateHoverEnd(date);
	};

	return (
		<td className={dateClassNames} role={disabled ? "presentation" : "gridcell"}>
			<Actionable
				fullWidth
				insetFocus
				className={s["cell-button"]}
				disabled={disabled}
				onClick={handleClick}
				attributes={{
					role: "checkbox",
					tabIndex: focusable ? 0 : -1,
					"aria-hidden": disabled,
					"aria-label": renderAriaLabel
						? renderAriaLabel({ date })
						: date.toLocaleDateString("en-us", { month: "long", day: "numeric", weekday: "long" }),
					"aria-checked": isActiveStart || isActiveEnd,
					"data-rs-date": getLocalISODate({ date }),
					onMouseEnter: handleMouseEnter,
					onMouseLeave: handleMouseLeave,
					onFocus: handleMouseEnter,
					onBlur: handleMouseLeave,
				}}
			>
				{date.getDate()}
			</Actionable>
		</td>
	);
};

export default CalendarDate;
