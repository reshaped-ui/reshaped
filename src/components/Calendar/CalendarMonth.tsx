"use client";

import { getMonthWeeks, getWeekdayNames, getLocalISODate } from "./Calendar.utils";
import CalendarDate from "./CalendarDate";
import type * as T from "./Calendar.types";
import s from "./Calendar.module.css";

const CalendarMonth = (props: T.MonthProps) => {
	const {
		date,
		value,
		onChange,
		min,
		max,
		range,
		firstWeekDay,
		selectedDates,
		hoveredDate,
		onDateHover,
		onDateHoverEnd,
		renderWeekDay,
		renderDateAriaLabel,
	} = props;
	let foundFocusableDate = false;
	const month = date.getMonth();
	const weeks = getMonthWeeks({ date, firstWeekDay });
	const weekdayNames = getWeekdayNames({ firstWeekDay, renderWeekDay });

	return (
		<table className={s.selection} role="grid">
			<thead aria-hidden="true">
				<tr>
					{weekdayNames.map((dayName) => (
						<th className={s.weekday} key={dayName} scope="col">
							{dayName}
						</th>
					))}
				</tr>
			</thead>
			<tbody>
				{weeks.map((week) => {
					const key = [month, week[0]?.getDate()].filter(Boolean).join("-");
					return (
						<tr key={key} className={s.row}>
							{week.map((date, index) => {
								const disabled = !!date && ((min && date < min) || (max && date > max));
								const month = date?.getMonth();
								const today = new Date();
								const isoToday = getLocalISODate({ date: today });
								const startValue = value && "start" in value ? value.start : value;
								const endValue = value && "end" in value ? value.end : value;
								const isoDate = date && getLocalISODate({ date });

								/**
								 * Decide if date has to be focusable with Tab (only one date should be)
								 * 1. If there is a selected value - it's focusable
								 * 2. Otherwise, today's date is focusable
								 * 3. Otherwise, first non-disabled date is focusable
								 */
								let focusable = false;

								if (!foundFocusableDate && date) {
									if (!!startValue && startValue.getMonth() === date?.getMonth()) {
										focusable = getLocalISODate({ date: startValue }) === getLocalISODate({ date });
									} else if (isoDate && month === today.getMonth()) {
										focusable = isoDate >= isoToday && !disabled;
									} else {
										focusable = !disabled;
									}
								}

								if (focusable) foundFocusableDate = true;

								return (
									<CalendarDate
										key={index}
										date={date}
										isoDate={isoDate}
										disabled={disabled}
										range={range}
										focusable={focusable}
										startValue={startValue}
										endValue={endValue}
										onChange={onChange}
										hoveredDate={hoveredDate}
										onDateHover={onDateHover}
										onDateHoverEnd={onDateHoverEnd}
										renderAriaLabel={renderDateAriaLabel}
										selectedDates={selectedDates}
									/>
								);
							})}
						</tr>
					);
				})}
			</tbody>
		</table>
	);
};

export default CalendarMonth;
