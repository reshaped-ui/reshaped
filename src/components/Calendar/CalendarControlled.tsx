"use client";

import React from "react";
import View from "components/View";
import { onNextFrame } from "utilities/animation";
import {
	setMonthToPrevious,
	setMonthToNext,
	setMonthTo,
	setYearToNext,
	setYearToPrevious,
	applyNavigationBounds,
} from "./Calendar.utils";
import useCalendarKeyboardNavigation from "./useCalendarKeyboardNavigation";
import CalendarMonth from "./CalendarMonth";
import CalendarYear from "./CalendarYear";
import CalendarControls from "./CalendarControls";
import type * as T from "./Calendar.types";

const CalendarControlled = (props: T.ControlledProps & T.BaseProps) => {
	const {
		value,
		onChange,
		defaultMonth,
		min,
		max,
		range,
		firstWeekDay,
		renderMonthLabel,
		renderSelectedMonthLabel,
		renderWeekDay,
		previousMonthAriaLabel,
		previousYearAriaLabel,
		nextMonthAriaLabel,
		nextYearAriaLabel,
		monthSelectionAriaLabel,
		renderMonthAriaLabel,
		renderDateAriaLabel,
	} = props;
	const [selectionMode, setSelectionMode] = React.useState<T.SelectionMode>("date");
	const [monthDate, setMonthDate] = React.useState(defaultMonth || new Date());
	const [hoveredDate, setHoveredDate] = React.useState<Date | null>(null);
	const monthTitleRef = React.useRef<HTMLButtonElement | null>(null);
	const prevSelectionModeRef = React.useRef<typeof selectionMode>(selectionMode);
	const bounds = applyNavigationBounds({ date: monthDate, min, max });
	const selectionRootRef = React.useRef<HTMLDivElement>(null);

	const handlePreviousClick = () => {
		if (selectionMode === "month") {
			setMonthDate((prev) => setYearToPrevious(prev));
			return;
		}

		setMonthDate((prev) => setMonthToPrevious(prev));
	};

	const handleNextClick = () => {
		if (selectionMode === "month") {
			setMonthDate((prev) => setYearToNext(prev));
			return;
		}

		setMonthDate((prev) => setMonthToNext(prev));
	};

	const handleMonthTitleClick = () => {
		setSelectionMode("month");
	};

	const handleMonthClick = (i: number) => {
		setMonthDate((prev) => setMonthTo(prev, i));
		setSelectionMode("date");
	};

	const handleDateHover = (date: Date) => {
		setHoveredDate(date);
	};

	const handleDateHoverEnd = (date: Date) => {
		if (hoveredDate && +hoveredDate === +date) setHoveredDate(null);
	};

	React.useEffect(() => {
		if (selectionMode === "date" && selectionMode !== prevSelectionModeRef.current) {
			onNextFrame(() => {
				monthTitleRef.current?.focus();
			});
		}

		prevSelectionModeRef.current = selectionMode;
	}, [selectionMode]);

	useCalendarKeyboardNavigation({
		monthDate,
		rootRef: selectionRootRef,
		changeToNextMonth: handleNextClick,
		changeToPreviousMonth: handlePreviousClick,
		// Each row has 7 days in date selection and 3 months in month
		verticalDelta: selectionMode === "date" ? 7 : 3,
		min,
		max,
	});

	return (
		<View gap={2}>
			<CalendarControls
				renderSelectedMonthLabel={renderSelectedMonthLabel}
				monthDate={monthDate}
				selectionMode={selectionMode}
				isFirstMonth={bounds.isFirstMonth}
				isLastMonth={bounds.isLastMonth}
				monthTitleRef={monthTitleRef}
				onMonthTitleClick={handleMonthTitleClick}
				onNextClick={handleNextClick}
				onPreviousClick={handlePreviousClick}
				previousMonthAriaLabel={previousMonthAriaLabel}
				previousYearAriaLabel={previousYearAriaLabel}
				nextMonthAriaLabel={nextMonthAriaLabel}
				nextYearAriaLabel={nextYearAriaLabel}
				monthSelectionAriaLabel={monthSelectionAriaLabel}
			/>

			<View.Item attributes={{ ref: selectionRootRef }}>
				{selectionMode === "date" && (
					<CalendarMonth
						date={monthDate}
						value={value}
						onChange={onChange}
						min={min}
						max={max}
						range={range}
						firstWeekDay={firstWeekDay}
						hoveredDate={hoveredDate}
						onDateHover={handleDateHover}
						onDateHoverEnd={handleDateHoverEnd}
						renderWeekDay={renderWeekDay}
						renderDateAriaLabel={renderDateAriaLabel}
					/>
				)}

				{selectionMode === "month" && (
					<CalendarYear
						monthDate={monthDate}
						onMonthClick={handleMonthClick}
						renderMonthLabel={renderMonthLabel}
						renderMonthAriaLabel={renderMonthAriaLabel}
						min={min}
						max={max}
					/>
				)}
			</View.Item>
		</View>
	);
};

export default CalendarControlled;
