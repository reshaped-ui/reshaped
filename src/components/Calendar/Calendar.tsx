import CalendarControlled from "./CalendarControlled";
import CalendarUncontrolled from "./CalendarUncontrolled";
import type * as T from "./Calendar.types";

const Calendar = (props: T.Props) => {
	if (props.value !== undefined) return <CalendarControlled {...props} />;
	return <CalendarUncontrolled {...props} />;
};

Calendar.displayName = "Calendar";

export default Calendar;
