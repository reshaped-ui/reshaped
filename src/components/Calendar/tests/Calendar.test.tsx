import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Calendar from "components/Calendar";
import Reshaped from "components/Reshaped";

beforeAll(() => {
	jest.useFakeTimers();
	jest.setSystemTime(new Date(2020, 0, 1));
});

afterAll(() => {
	jest.useRealTimers();
});

describe("Components/Calendar", () => {
	test("renders calendar for current month", () => {
		render(
			<Reshaped>
				<Calendar />
			</Reshaped>
		);

		const monthTitleEl = screen.getByText("January 2020");

		const gridEl = screen.getByRole("grid");
		const gridCellEls = screen.getAllByRole("gridcell");
		const dateEls = screen.getAllByRole("checkbox");
		const buttonEls = screen.getAllByRole("button");

		expect(monthTitleEl).toBeInTheDocument();
		expect(gridEl).toBeInTheDocument();
		expect(gridCellEls).toHaveLength(31);

		expect(dateEls).toHaveLength(31);
		expect(dateEls[0].getAttribute("aria-label")).toMatchSnapshot();
		expect(dateEls[0]).toHaveAttribute("tabIndex", "0");
		expect(dateEls[1]).toHaveAttribute("tabIndex", "-1");

		// Previous month button
		expect(buttonEls[0].getAttribute("aria-label")).toMatchSnapshot();
		// Month select button
		expect(buttonEls[1].innerHTML).toMatchSnapshot();
		// Next month button
		expect(buttonEls[2].getAttribute("aria-label")).toMatchSnapshot();
	});

	test("renders passed month", () => {
		render(
			<Reshaped>
				<Calendar defaultMonth={new Date(2022, 11)} />
			</Reshaped>
		);

		const monthTitleEl = screen.getByText("December 2022");

		expect(monthTitleEl).toBeInTheDocument();
	});

	test("supports onChange", async () => {
		const handleChange = jest.fn();

		render(
			<Reshaped>
				<Calendar onChange={handleChange} />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");

		await userEvent.click(dateEls[0], { delay: null });

		await waitFor(() => {
			expect(handleChange).toHaveBeenCalledTimes(1);
			expect(handleChange).toHaveBeenCalledWith({ value: new Date(2020, 0, 1) });
		});
	});

	test("supports single defaultValue", async () => {
		render(
			<Reshaped>
				<Calendar defaultValue={new Date(2020, 0, 5)} />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");
		const selectedDateEl = dateEls[4];

		expect(selectedDateEl).toBeInTheDocument();
		expect(selectedDateEl).toHaveAttribute("tabIndex", "0");
		expect(selectedDateEl).toHaveAttribute("aria-checked", "true");

		await userEvent.click(dateEls[6], { delay: null });

		await waitFor(() => {
			expect(dateEls[6]).toHaveAttribute("aria-checked", "true");
		});
	});

	test("supports single value", async () => {
		render(
			<Reshaped>
				<Calendar value={new Date(2020, 0, 5)} />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");
		const selectedDateEl = dateEls[4];

		expect(selectedDateEl).toBeInTheDocument();
		expect(selectedDateEl).toHaveAttribute("tabIndex", "0");
		expect(selectedDateEl).toHaveAttribute("aria-checked", "true");

		await userEvent.click(dateEls[6], { delay: null });

		await waitFor(() => {
			expect(dateEls[6]).toHaveAttribute("aria-checked", "false");
		});
	});

	test("supports range defaultValue", async () => {
		render(
			<Reshaped>
				<Calendar range defaultValue={{ start: new Date(2020, 0, 5), end: new Date(2020, 0, 7) }} />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");
		const startDateEl = dateEls[4];
		const endDateEl = dateEls[6];

		expect(startDateEl).toHaveAttribute("tabIndex", "0");
		expect(startDateEl).toHaveAttribute("aria-checked", "true");

		expect(endDateEl).toHaveAttribute("tabIndex", "-1");
		expect(endDateEl).toHaveAttribute("aria-checked", "true");

		await userEvent.click(dateEls[3], { delay: null });

		await waitFor(() => {
			expect(dateEls[3]).toHaveAttribute("aria-checked", "true");
			expect(dateEls[3]).toHaveAttribute("tabIndex", "0");

			expect(startDateEl).toHaveAttribute("aria-checked", "false");
			expect(startDateEl).toHaveAttribute("tabIndex", "-1");

			expect(endDateEl).toHaveAttribute("aria-checked", "false");
			expect(endDateEl).toHaveAttribute("tabIndex", "-1");
		});

		// Click before the start date
		await userEvent.click(dateEls[2], { delay: null });

		await waitFor(() => {
			expect(dateEls[2]).toHaveAttribute("aria-checked", "true");
			expect(dateEls[2]).toHaveAttribute("tabIndex", "0");

			expect(dateEls[3]).toHaveAttribute("aria-checked", "false");
			expect(dateEls[3]).toHaveAttribute("tabIndex", "-1");
		});

		await userEvent.click(dateEls[4], { delay: null });

		await waitFor(() => {
			expect(dateEls[2]).toHaveAttribute("aria-checked", "true");
			expect(dateEls[2]).toHaveAttribute("tabIndex", "0");

			expect(dateEls[4]).toHaveAttribute("aria-checked", "true");
			expect(dateEls[4]).toHaveAttribute("tabIndex", "-1");
		});
	});

	test("supports min/max boundaries", () => {
		render(
			<Reshaped>
				<Calendar min={new Date(2020, 0, 5)} max={new Date(2020, 0, 25)} />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");

		expect(dateEls).toHaveLength(21);
	});

	test("supports month selection", async () => {
		render(
			<Reshaped>
				<Calendar />
			</Reshaped>
		);

		const buttonEls = screen.getAllByRole("button");

		await userEvent.click(buttonEls[1], { delay: null });

		const monthEls = screen.getAllByRole("button").filter((el) => el.hasAttribute("data-rs-date"));

		expect(monthEls).toHaveLength(12);
	});

	test("supports custom aria labels", async () => {
		render(
			<Reshaped>
				<Calendar
					nextYearAriaLabel="Test next year"
					nextMonthAriaLabel="Test next month"
					renderDateAriaLabel={() => "Test date"}
					renderMonthAriaLabel={() => "Test month"}
					previousYearAriaLabel="Test previous year"
					previousMonthAriaLabel="Test previous month"
					monthSelectionAriaLabel="Test month selection"
				/>
			</Reshaped>
		);

		const buttonEls = screen.getAllByRole("button");
		const dateEls = screen.getAllByRole("checkbox");

		expect(buttonEls[0]).toHaveAttribute("aria-label", "Test previous month");
		expect(buttonEls[1]).toHaveTextContent("Test month selection");
		expect(buttonEls[2]).toHaveAttribute("aria-label", "Test next month");
		expect(dateEls[0]).toHaveAttribute("aria-label", "Test date");

		await userEvent.click(buttonEls[1], { delay: null });

		const yearButtonEls = screen.getAllByRole("button");

		expect(yearButtonEls[0]).toHaveAttribute("aria-label", "Test previous year");
		expect(yearButtonEls[1]).toHaveAttribute("aria-label", "Test next year");
		expect(yearButtonEls[2]).toHaveAttribute("aria-label", "Test month");
	});

	test("supports keyboard navigation", async () => {
		render(
			<Reshaped>
				<Calendar />
			</Reshaped>
		);

		const dateEls = screen.getAllByRole("checkbox");

		act(() => {
			dateEls[0].focus();
		});

		await userEvent.keyboard("{ArrowRight}", { delay: null });
		expect(document.activeElement).toHaveAttribute("data-rs-date", "2020-01-02");

		await userEvent.keyboard("{ArrowLeft}", { delay: null });
		expect(document.activeElement).toHaveAttribute("data-rs-date", "2020-01-01");

		await userEvent.keyboard("{ArrowDown}", { delay: null });
		expect(document.activeElement).toHaveAttribute("data-rs-date", "2020-01-08");

		await userEvent.keyboard("{ArrowUp}", { delay: null });
		expect(document.activeElement).toHaveAttribute("data-rs-date", "2020-01-01");

		await userEvent.keyboard("{ArrowUp}", { delay: null });
		expect(document.activeElement).toHaveAttribute("data-rs-date", "2019-12-25");
	});
});
