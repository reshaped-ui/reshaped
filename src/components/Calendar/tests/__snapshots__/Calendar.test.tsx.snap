// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Calendar renders calendar for current month 1`] = `"Wednesday, January 1"`;

exports[`Components/Calendar renders calendar for current month 2`] = `"Previous month"`;

exports[`Components/Calendar renders calendar for current month 3`] = `"<span class="text">January 2020<div class="root">Select a month</div></span>"`;

exports[`Components/Calendar renders calendar for current month 4`] = `"Next month"`;
